ably/realtime_channel_spec_test.go

```diff
@@ -155,6 +155,662 @@ func attachAndWait(t *testing.T, channel *ably.RealtimeChannel) {
+		afterCalls = make(chan ablytest.AfterCall, 1)
+		now, after := ablytest.TimeFuncs(afterCalls)
+
+		c, _ = ably.NewRealtime(
+			ably.WithToken("fake:token"),
+			ably.WithAutoConnect(false),
+			ably.WithNow(now),
+			ably.WithAfter(after),
+			ably.WithChannelRetryTimeout(channelRetryTimeout),
+			ably.WithRealtimeRequestTimeout(realtimeRequestTimeout),
+			ably.WithDial(ablytest.MessagePipe(in, out)),
+		)
+
+		in <- &proto.ProtocolMessage{
+			Action:            proto.ActionConnected,
+			ConnectionID:      "connection-id",
+			ConnectionDetails: &proto.ConnectionDetails{},
+		}
+
+		err := ablytest.Wait(ablytest.ConnWaiter(c, c.Connect, ably.ConnectionEventConnected), nil)
+		if err != nil {
+			t.Fatal(err)
+		}
+
+		channel = c.Channels.Get("test")
+		stateChanges = make(ably.ChannelStateChanges, 10)
+		return
+	}
+
+	t.Run("RTL5a", func(t *testing.T) {
+		_, out, _, channel, stateChanges, _ := setup(t)
+		ctx, cancel := context.WithCancel(context.Background())
+		cancel()
+		channel.OnAll(stateChanges.Receive)
+
+		// Initialized state
+		channel.Detach(ctx)
+		ablytest.Instantly.NoRecv(t, nil, out, t.Fatalf)          // Should not send detach message
+		ablytest.Instantly.NoRecv(t, nil, stateChanges, t.Fatalf) // Should not make any change to the state
+
+		channel.SetState(ably.ChannelStateDetached, nil)
+		ablytest.Instantly.Recv(t, nil, stateChanges, t.Fatalf) // State will be changed to detached
+
+		// Detached state
+		channel.Detach(ctx)
+		ablytest.Instantly.NoRecv(t, nil, out, t.Fatalf)          // Should not send detach message
+		ablytest.Instantly.NoRecv(t, nil, stateChanges, t.Fatalf) // Should not make any change to the state
+	})
+
+	t.Run("RTL5b", func(t *testing.T) {
+		_, _, _, channel, stateChanges, _ := setup(t)
+		ctx, cancel := context.WithCancel(context.Background())
+		cancel()
+		channel.OnAll(stateChanges.Receive)
+
+		channel.SetState(ably.ChannelStateFailed, nil)
+		ablytest.Instantly.Recv(t, nil, stateChanges, t.Fatalf) // State will be changed to failed
+		err := channel.Detach(ctx)
+
+		if expected, got := "cannot Detach channel because it is in FAILED state", err.Error(); !strings.Contains(got, expected) {
+			t.Fatalf("expected error %+v; got %v", expected, got)
+		}
+	})
+
+	t.Run("RTL5d RTL5e", func(t *testing.T) {
+		in, out, _, channel, stateChanges, _ := setup(t)
+		ctx, cancel := context.WithCancel(context.Background())
+		defer cancel()
+		channel.OnAll(stateChanges.Receive)
+
+		// Get the channel to ATTACHED.
+
+		in <- &proto.ProtocolMessage{
+			Action:  proto.ActionAttached,
+			Channel: channel.Name,
+		}
+
+		var change ably.ChannelStateChange
+
+		ablytest.Instantly.Recv(t, &change, stateChanges, t.Fatalf)
+		if expected, got := ably.ChannelStateAttached, change.Current; expected != got {
+			t.Fatalf("expected %v; got %v (event: %+v)", expected, got, change)
+		}
+
+		var outMsg *proto.ProtocolMessage
+
+		go func() {
+			time.Sleep(time.Second / 2)
```

Comment:
What's the point of waiting? This only makes the test flaky, while you control all events anyway since the connection transport is mocked. You can just do this below the DETACHING check synchronously.